fix #3960 adr(visualization): Add ADR for monitoring data import.

emtwo opened the pull request.

jaredlockhart commented on app/experimenter/docs/adrs/0004-monitoring-data-import.md:
```diff
@@ -0,0 +1,67 @@
+1. Add ETL in [biguqery_etl](https://github.com/mozilla/bigquery-etl) to process the live data periodically and output results to a gcs bucket that can be accessed by Experimenter. This is how Jetstream data is currently being accessed.
+2. Add a background celery task that fetches data on-demand from bigquery and stores it in Redis. Redis cache can be used for the next fetch of the same data, but it gets invalidated periodically.
+3. Embed existing Grafana graphs (e.g. [graphs](https://grafana.telemetry.mozilla.org/d/XspgvdxZz/experiment-enrollment?orgId=1&var-experiment_id=bug-1656561-rapid-testing-demo-nightly-80&from=1596153600000&to=1598832000000))
+
```
I'd like to know if there's a way to do it by running rolling accumulators for enrollment/unenrollment events at the pubsub level that could compute continuous population levels for every live experiment that would give us an immediate sense of every live experiments enrollment without doing it at the bq level, either on demand or scheduled etl?

sunahsuh replied:
When I was looking at this earlier, reading the event pings from pubsub would be prohibitively expensive – this might change if/when we move to pubsub lite, but we'd need to do some analysis on costs. The other mark against this is that it's a special case that would increase maintenance burden vs using existing, established tooling.

jaredlockhart replied:
Ah that's too bad, then if it has to be done at the bq level then @emtwo 's right we should just do it on demand with a celery task 👍

jaredlockhart reviewed (commented):
Great writeup @emtwo !

sunahsuh commented on app/experimenter/docs/adrs/0004-monitoring-data-import.md:
```diff
@@ -0,0 +1,67 @@
+
+1. Add ETL in [biguqery_etl](https://github.com/mozilla/bigquery-etl) to process the live data periodically and output results to a gcs bucket that can be accessed by Experimenter. This is how Jetstream data is currently being accessed.
+2. Add a background celery task that fetches data on-demand from bigquery and stores it in Redis. Redis cache can be used for the next fetch of the same data, but it gets invalidated periodically.
+3. Embed existing Grafana graphs (e.g. [graphs](https://grafana.telemetry.mozilla.org/d/XspgvdxZz/experiment-enrollment?orgId=1&var-experiment_id=bug-1656561-rapid-testing-demo-nightly-80&from=1596153600000&to=1598832000000))
```
Side questions that came up as I was thinking about this: does this mean that experimenter would have fixed views for the data? that is, will users be able to control the time range that's graphed? Because in that case fixed json would probably not work.

Also, have we considered how we'd handle long-running experiments? Because, grafana solves there's some sticky issues around the number of datapoints to retrieve based on the timeframe that's being graphed (i.e. if we're looking at a graph over several months, the query that grafana issues does not return one row per 5 minute aggregation window but is instead scaled appropriately.)

emtwo replied:
> Side questions that came up as I was thinking about this: does this mean that experimenter would have fixed views for the data? that is, will users be able to control the time range that's graphed? Because in that case fixed json would probably not work.

@sunahsuh could you elaborate on why this wouldn't work? Is it too much data to be storing in a bucket and deciding client-side what range is shown?

> Also, have we considered how we'd handle long-running experiments? Because, grafana solves there's some sticky issues around the number of datapoints to retrieve based on the timeframe that's being graphed (i.e. if we're looking at a graph over several months, the query that grafana issues does not return one row per 5 minute aggregation window but is instead scaled appropriately.)

I'm guessing this is related to the question above in that, we cannot place 5-minute aggregations in a bucket for several months. Perhaps the job that runs every 5 min to output to buckets will also scale the aggregation depending on how much data is available? E.g. if more than 1 day of data is available, it will no longer continue to store 5-minute intervals but daily intervals in the bucket.

scholtzan replied:
The cost for updating the hourly table per run is ~20GB per run, so this might even put us around $1000 for the monthly cost with option 1.

I'm guessing exporting data to GCS will mean one json file per experiment? 

Also, as an alternative to sending all data to the client and filtering it there based on the data range, perhaps Experimenter could query experiments enrollments table directly with the appropriate date range and could aggregate data if that range exceeds some threshold? This would reduce the amount of data sent to the client, which can get quite large over time with 5 minute intervals for multiple branches. At the same time it would still be possible to let users zoom into the graph to get more detailed "5 minute interval" information.

emtwo replied:
> The cost for updating the hourly table per run is ~20GB per run, so this might even put us around $1000 for the monthly cost with option 1.

Do you have links for the queries you tested with? And maybe also a breakdown of how you got to the $1000 estimate. I just want to understand more clearly if that's just the cost to update the hourly table or also the cost to run the 5-minute job to query the union-ed daily/hourly/live tables.

I think it would be helpful to see how the cost per run with vs. without the hourly table included in the union would be, I'm assuming this is where we would see the most cost gains.

> I'm guessing exporting data to GCS will mean one json file per experiment?

Yes, I would say so.

> Also, as an alternative to sending all data to the client and filtering it there based on the data range, perhaps Experimenter could query experiments enrollments table directly with the appropriate date range and could aggregate data if that range exceeds some threshold? This would reduce the amount of data sent to the client, which can get quite large over time with 5 minute intervals for multiple branches. At the same time it would still be possible to let users zoom into the graph to get more detailed "5 minute interval" information.

Potentially we could do this, yes. I think whether queries need to happen in experimenter or not is the big question I have because a bunch of extra infrastructure needs to be added to experimenter to query BQ directly from there. So I'd love to know, for example, if custom date/time ranges is not a requirement, would that mean we could rely entirely on the output in the buckets created by the 5-minute interval job?

scholtzan replied:
The query I tested is this one, which is basically the current query for enrollment aggregates: 
https://console.cloud.google.com/bigquery?sq=1018637472215:70c8f25a689d42f5a09f3cffe95fe401

Note that the dryrun numbers are incorrect and much higher than what is actually processed.

My estimate:
For the hourly job: (20GB * 24hours * 30days) / 1024GB * 5$ = 70$/month

For the 5-minute job, which will query the live table (up to 1.5h of data that is not in the hourly table) and the hourly table: (30GB * 12 * 24h * 30days) / 1024GB * 5$ = 1265$/month
The 30GB here is an upper limit, for when the past 1.5h are queried. It'll be below that for most runs.

scholtzan replied:
Having a fixed date range would certainly simplify things and we could potentially just use the output in the bucket. We might need to aggregate data (maybe 10 minute intervals instead of 5, or hourly, ...) if clients can't handle processing the amount of data that is sent. Since there might also be multiple visualizations (for daily active population, daily active population by branch, cumulative population estimate, cumulative population estimate by branch, ...) I suppose.

emtwo replied:
Cool! This is helpful, thank you!

> The query I tested is this one, which is basically the current query for enrollment aggregates:
> https://console.cloud.google.com/bigquery?sq=1018637472215:70c8f25a689d42f5a09f3cffe95fe401

I want to clarify here about my interpretation of Sunah's suggestion which seems maybe different than the approach here and maybe also not actually what Sunah suggested lol. I think I may need to write up something more detailed about this for clarity so we can all be on the same page.

So my understanding was that the hourly job would look almost exactly like the daily job that runs this: https://github.com/mozilla/bigquery-etl/blob/master/sql/moz-fx-data-shared-prod/telemetry_derived/experiment_enrollment_aggregates_v1/query.sql Except of course the filter would be for 1 hour instead of 1 day. It would append to a table once per hour with the 5-min interval aggregations.

So now we would have a new table that would need to be included in this view, right here, as a third table in the union for `all_enrollments`: https://github.com/mozilla/bigquery-etl/blob/master/sql/moz-fx-data-shared-prod/telemetry_derived/experiment_enrollment_aggregates_live/view.sql.py#L80

Right now, that view, I believe, is only processed on demand when it's queried by grafana. So the 5-minute job, like you calculated above would be querying this view but I *think* the cost might be a tad higher when the additional unioned table is added.

Additionally, like you said, @scholtzan  the 5 minute job would need to include all the grafana queries, which I think will really add up and will reduce flexibility for experimenter to allow users to set time ranges. But it would be useful to have the cost estimates for these anyway. Though I'm starting to get the feeling we will need to be making these queries from experimenter.

@sunahsuh wanted to confirm this is an accurate description of what you meant? And if so, then we need the estimate for the hourly table as well as the 5-min job with the newly added unioned table. Thanks!

sunahsuh replied:
> I want to clarify here about my interpretation of Sunah's suggestion which seems maybe different than the approach here and maybe also not actually what Sunah suggested lol.

This description sounds right to me!

> Additionally, like you said, @scholtzan the 5 minute job would need to include all the grafana queries, which I think will really add up and will reduce flexibility for experimenter to allow users to set time ranges. But it would be useful to have the cost estimates for these anyway.

I think we could potentially create the json with a single query that pulls data for every graph for every experiment if we do a lot of the munging in the output job (meaning this will probably be a custom python script.)

One thing to clarify is that I'm not as concerned with the size of the json blob so much as the complexity that we're taking on, either on the client side or in the json output job (depending on whether we allow fixed time ranged or not), because we'll need to do time-scaled aggregation to create a decent user experience for them. Consider that one datapoint every five minutes translates to 288 datapoints per day, meaning after a week any given graph will have over 2000 datapoints. To graph that in a usable way we need to aggregate datapoints. This is something that Grafana takes care of under the hood at the query level (hence why the time select in a Grafana query looks like: `timestamp_millis(cast((unix_millis(window_start)/$__interval_ms) as int64)*$__interval_ms) AS time`, where `$__interval_ms` is a Grafana magic variable that it fills in based on the time period being graphed.) It might be somewhat invisible but Grafana does handle more complexity than it might appear at first glance because it's so well-tuned for displaying time series dashboards for the purpose of monitoring. These are things we'd need to now handle ourselves one way or the other. I'm not sure what capabilities the graphing libraries we're considering might have built-in but we do always have the option of using grafana embedded graphs with a custom theme if we want to ship faster and push this issue off until later, and tackle driving down costs separately.

scholtzan replied:
For some additional clarification: 

> So my understanding was that the hourly job would look almost exactly like the daily job that runs this: https://github.com/mozilla/bigquery-etl/blob/master/sql/moz-fx-data-shared-prod/telemetry_derived/experiment_enrollment_aggregates_v1/query.sql Except of course the filter would be for 1 hour instead of 1 day. It would append to a table once per hour with the 5-min interval aggregations.

It would look similar to https://github.com/mozilla/bigquery-etl/blob/master/sql/moz-fx-data-shared-prod/telemetry_derived/experiment_enrollment_aggregates_v1/query.sql, but instead would operate on the live table instead of `telemetry_derived.event_events_v1`.

> So now we would have a new table that would need to be included in this view, right here, as a third table in the union for all_enrollments: https://github.com/mozilla/bigquery-etl/blob/master/sql/moz-fx-data-shared-prod/telemetry_derived/experiment_enrollment_aggregates_live/view.sql.py#L80

Yes. If we have the hourly table, then we can reduce the time range for `live` to recent data that has not been processed by the hourly job yet. Which will make it cheaper.

I opened a non-public ticket with some updated cost numbers here: https://bugzilla.mozilla.org/show_bug.cgi?id=1680319

I'm not sure how much value custom date ranges will actually have for users compared to just showing data for the entire experiment period and maybe thinking of a good approach for aggregating data (1 hour intervals, daily intervals, ...) if that period is larger than some threshold.

sunahsuh replied:
Looks like Anna and my comments may have collided :D

> I'm not sure how much value custom date ranges will actually have for users compared to just showing data for the entire experiment period and maybe thinking of a good approach for aggregating data (1 hour intervals, daily intervals, ...) if that period is larger than some threshold.

Yeah I think I'd lean toward fixed time ranges, and if we do that, then I suppose we can figure out a formula for granularity and specify the aggregations we'd want to see. And if people want to explore things in more detail I don't think it'd be too hard to create an external dashboard in $future_bi_tool they can use as a starting point.

emtwo replied:
> I'm not sure what capabilities the graphing libraries we're considering might have built-in but we do always have the option of using grafana embedded graphs with a custom theme if we want to ship faster and push this issue off until later, and tackle driving down costs separately.

I agree with you both @sunahsuh and @scholtzan about fixed time ranges. I also think that between needing a python script for the grafana query replacements and needing to address time scales as data rolls in, this is quite a lot of work that I don't want to block on to have some basic monitoring in experimenter.

I think what I'll do is summarize as best as I can the discussion here and add it as a 4th option. The new approach I want to propose is to do option 2 in the near term for only some small amount of monitoring data (we really need basic enrollment counts and health monitoring) We can link out to grafana or embed the other graphs for now. Then do the 4th option addressing costs as a longer term solution. How does that sound?

I'll write this out and update this doc.

scholtzan replied:
That sounds ok to me. Depending on how much effort it is to implement option 2 and migrate to the long-term solution later, maybe it is also enough to just link to the Grafana dashboards for now?

Would it make sense to write a proposal for the long-term solution? Might be good to clarify requirements and define how data should be stored for Experimenter to consume it etc. Also just thinking through what queries and parts are required would be good. I'd be happy to write the implementation/ETL part of the proposal.

emtwo replied:
> That sounds ok to me. Depending on how much effort it is to implement option 2 and migrate to the long-term solution later, maybe it is also enough to just link to the Grafana dashboards for now?

I think it won't be *too* bad to implement #2 and later swap it out in experimenter to query GCS instead of BQ directly. I think it should take 1-2 weeks of work or so to implement it then later swap it. I suspect between the ETL proposal + work it would probably take longer than a couple weeks so I'm not sure about waiting. Also if we have this intermediate solution working, we don't have to worry about rushing to get all the ETL set up and we can take our time to get it right. 

> Would it make sense to write a proposal for the long-term solution? Might be good to clarify requirements and define how data should be stored for Experimenter to consume it etc. Also just thinking through what queries and parts are required would be good. I'd be happy to write the implementation/ETL part of the proposal.

Yes, I think the long term solution does need its own proposal and it seems you've already done a bunch of the upfront work to be able to write that up! ;)

scholtzan replied:
Yeah, I'd expect the long-term solution to be available early Q1 2021.

sunahsuh reviewed (commented):
I may have brought up more questions than answers as I was describing the hourly thing

scholtzan commented:
I did some investigation in using a [materialized view](https://cloud.google.com/bigquery/docs/materialized-views) for experiments enrollments. Materialized views have quite a few [limitations](https://cloud.google.com/bigquery/docs/materialized-views-intro#limitations) at the moment, such as 
_A materialized view is limited to referencing only a single table, and is not able to use joins or UNNEST functionality_. The experiment enrollments query needs to access event tables and retrieve branch information from`event_map_values`. However, with `event_map_values` being an array, it's currently not possible to use materialized views.

pdehaan commented on app/experimenter/docs/adrs/0004-monitoring-data-import.md:
```diff
@@ -0,0 +1,67 @@
+* Pros:
+  * It won't be any slower than the current Grafana graphs
+  * Allows flexibility for how data is visualized in Experimenter
+  * The caching layer would make this approach the same price or cheaper than (1) and (2)
```
Should this say "(1) and (3)"? (Since this is option (2))

emtwo replied:
yes, thank you

emtwo commented:
@sunahsuh @scholtzan I've updated with a summary of the discussion above and explained it's a long term option that needs further breakdown. I've also added your names at the top of the doc as contributors. Thanks so much for your help!

I added you for final review.

sunahsuh commented on app/experimenter/docs/adrs/0004-monitoring-data-import.md:
```diff
@@ -0,0 +1,92 @@
+#### (4) Option (1) + ETL Optimizations
+Option (1) would be the ideal solution if it was not the most expensive. This ADR initiated ideas about ETL optimizations for option (1) to be implemented with much lower costs.
+
+The idea is to add an hourly job that looks almost exactly like the [daily job](https://github.com/mozilla/bigquery-etl/blob/master/sql/moz-fx-data-shared-prod/telemetry_derived/experiment_enrollment_aggregates_v1/query.sql) that already runs to aggregate 5-minute intervals of live data. Instead, however, the filter would be for 1 hour instead of 1 day and it would operate on the live table inste ...
```
Should we mention the lag here? I'm looking for the docs where this lives, but IIRC the pipeline's SLA is 15 minutes in lag for processing

sunahsuh reviewed (approved):
Looks great, thank you for an interesting discussion, all!
